File: heartbeat/heartbeat.go
package main
import (
    riffle "github.com/exis-io/core/riffle"
    validator "./validator"
    "time"
)

func hb_driver(){
    c := riffle.NewDomain("xs")
    sender := c.Subdomain("node")
    sender.Join()
        for {
            if validator.Hb_send {
                sender.Publish("can", "5C0#1301") //May want to send heartbeat can message here
                riffle.Info("Sending WCM heartbeat")
                time.Sleep(time.Duration(validator.Interval_ms) * time.Millisecond)
            }
        }
}

func hb_ctrl(data []int){

        if data[0] == 1 {
            riffle.Info("received hb start command")
            validator.Hb_send = true
            validator.Interval_ms = data[1]
        } else if (data[0] == 0) {
            riffle.Info("received hb stop command")
            validator.Hb_send = false
        }
}

func hb_handler(data []string){
    

    if validator.Rx_hb_empty {
        validator.StartTime = time.Now().UnixNano()%1e6/1e3
        validator.Rx_hb_empty = false
    }

    validator.Rx_hb[data[1]] = data[2]
    full:=true

    for _, v := range validator.Rx_hb { 
        if v == ""{
            full = false
        }
    }

    if full {
        for k, _ := range validator.Rx_hb { 
            validator.Rx_hb[k] = ""
        }
        validator.EndTime = time.Now().UnixNano()%1e6/1e3
        validator.Rx_hb_empty = true
    }

    if ((validator.EndTime - validator.StartTime)>int64(validator.Interval_ms)){
        validator.FaultCount = validator.FaultCount + 1
    }else{
        validator.FaultCount = 0
    }
}

func listen(app riffle.Domain){
        app.Listen()
}

func main() {
        riffle.SetLogLevelInfo()
        // Create a domain
        riffle.SetFabric("ws://192.168.1.99:9000")
        // Create the domain objects
        core := riffle.NewDomain("xs")
        app := core.Subdomain("node")
        app.Join()

        if e := app.Subscribe("hb_ctrl", hb_ctrl); e != nil {
                riffle.Info("Unable to subscribe to hb_ctrl endpoint", e.Error())
        } else {
                riffle.Info("Subscribed to hb_ctrl endpoint!")
        }

        if e := app.Subscribe("hb", hb_handler); e != nil {
                riffle.Info("Unable to subscribe to can endpoint", e.Error())
        } else {
                riffle.Info("Subscribed to heartbeat endpoint!")
        }
        // go hb_driver()
        listen(app)
        defer func(){listen(app)}()
}

File: heartbeat/heartbeat.py
import riffle
from riffle import want
import time
import os
import subprocess
import argparse
import json
import re

parser = argparse.ArgumentParser(description="A script for converting raw can messages to correct values")
parser.add_argument('-p','--parser',default='../dashboard/app/parser.json',help="Location of parser json file",metavar="parser")
parser.add_argument('-l','--backend_location',default="ws://localhost:8000", help="Location of backend.  Defaults to ws://localhost:8000",metavar="backend")
args = vars(parser.parse_args())

with open(args['parser']) as parser_file:    
    parser = json.load(parser_file)

current_milli_time = lambda: int(round(time.time() * 1000))

class Heartbeat():
    def __init__(self):
        self.start_time = None
        self.end_time = None
        self.fault_count = 0
        self.fault_max = 0
        self.interval_ms = 1000
        self.rx_hb_empty = True
        self.timer = time.time()
        self.modules = {
                'VSM' :{'prev_fault':0, 'fault': 0,'prev': 1, 'cur':1, 'next': 1, 'last_updated':time.time()},
                'BCM' :{'prev_fault':0, 'fault': 0,'prev': 1, 'cur':1, 'next': 1, 'last_updated':time.time()},
                'MCM' :{'prev_fault':0, 'fault': 0,'prev': 1, 'cur':1, 'next': 1, 'last_updated':time.time()},
                'WCM' :{'prev_fault':0, 'fault': 0,'prev': 1, 'cur':1, 'next': 1, 'last_updated':time.time()},
                'VNM' :{'prev_fault':0, 'fault': 0,'prev': 1, 'cur':1, 'next': 1, 'last_updated':time.time()},
                'BMS' :{'prev_fault':0, 'fault': 0,'prev': 1, 'cur':1, 'next': 1, 'last_updated':time.time()}
        }

    def check_timing_fault(self):
        now = time.time()
        for k in self.modules:
            time_delta = now - self.modules[k]['last_updated']
            if time_delta > self.interval_ms:
                self.fault_count = self.fault_count + 1
        self.fault_count = 0


    def generate_status(self):
        print("generating status")
        if self.fault_count>self.fault_max:
            return "FAILURE"
        for k in self.modules.keys():
            if self.modules[k]['fault'] is not 0 or self.modules[k]['fault'] in [1,6]:
                return "FAILURE"
        return "GOOD"


          
    def update_module_hb(self,module,data):
        self.modules[module]['last_updated'] = time.time()
        self.modules[module]['prev_fault']= int(data[0],16)
        self.modules[module]['fault']= int(data[1],16)
        self.modules[module]['prev']= int(data[2],16)
        self.modules[module]['cur']= int(data[3],16)
        self.modules[module]['next']= int(data[4],16)

class HB(riffle.Domain):

    def onJoin(self):
        print("Connected to Exis Node")
        self.Heartbeat = Heartbeat()
        self.subscribe("hb_ctrl", self.hb_ctrl)
        # self.subscribe("hb", self.hb_handler)
        self.subscribe("can", self.can_parser)
        self.p=None

    def send_hb(self,sid,data):
        print("updating heartbeat sid: " + str(sid))

        split_data = re.findall('..',data[3])
        # print(split_data)
        send = {}
        for module in parser['SID']:

            if 'from' in parser['SID'][module]:
                from_mask = parser['SID'][module]['from']
                # TODO Fix this logic
                and_res = int(sid,16) & from_mask
                # print(and_res)
                if and_res == from_mask and and_res != 0:
                    print("Got status for " + module)
                    self.Heartbeat.update_module_hb(module,split_data)
                    self.Heartbeat.check_timing_fault()
                    
                    send['modules'] = self.Heartbeat.modules
                    send['fault_count'] = self.Heartbeat.fault_count
                    send['system_status'] = self.Heartbeat.generate_status()
                    #print(send)
        if 'system_status' in send and (time.time() - self.Heartbeat.timer) >= self.Heartbeat.interval_ms:
            print('sending heartbeat')
            self.publish("hb",send)

    def can_parser(self,data):
        #TODO: Ensure this logic is correct
        converted_batch = []
        #print(data)
        # msg in format [timestamp,sid,msg_type,data]
        for msg in data:
            
            ts = float(msg[0])
            sid = msg[1]
            msg_type = int(msg[2],16)
            data_str = msg[3]
            if msg_type == 1:
                self.send_hb(sid,msg)
            else:
                converted_data = [ts,sid,msg_type]
                message_spec = parser['messages'][msg_type]
                for val in message_spec['values']:
                    data_value = data_str[:(val['byte_size']*2)]
                    data_str.replace(data_value,'',1)
                    # print(data_value)
                    if val['units'] in ['int','count','status','state']:
                        formatted_val = int(data_value,16)
                    else:
                        formatted_val = round(int(data_value,16)*val['scalar'],val['precision'])
                    # print(formatted_val)
                    #converted_data.append(formatted_val)
                    # if 'STRIP' in val['title']:
                    #     print(val['title'] + ": " + str(int(data_value,16)))
                    converted_batch.append([val['title'],formatted_val])
                # print("converting data")
                #converted_batch.append(converted_data)
        #print(converted_batch)
        print('Sending formatted data')
        self.publish("data",converted_batch)

    def hb_ctrl(self, data):
        if data[0] == 1:
            self.Heartbeat.fault_max = data[2]
            self.Heartbeat.interval_ms = data[1]
            self.Heartbeat.run_hb = True
            #Add more logic to check if hb is still running
            print("starting hb subprocess")
            self.p = subprocess.Popen("python hb-driver.py %s %s" %(data[1],args['backend_location']), stdout=subprocess.PIPE, shell=True)
        elif data[0] == 0:
            if self.p:
                self.p.kill()

    # def hb_handler(self, data):
    #     print("Received hb CAN message %s" %data)
    #     if self.Heartbeat.rx_hb_empty:
    #         self.Heartbeat.rx_hb[data[3]]



if __name__ == '__main__':
    #riffle.SetLogLevelDebug()
    riffle.SetFabric(args['backend_location'])
    HB('xs.node').join()

